part 1 transformer ranking (#84)
vespa-engine/blog: _posts/2021-05-19-pretrained-transformer-language-models-for-search-part-1.md

---
layout: post
title: "Pretrained Transformer Language Models for Search - part 1"
date: '2021-05-19'
tags: []
author: jobergum 
image: assets/2021-05-19-pretrained-transformer-language-models-for-search-part-1/jamie-street-uNNCs5kL70Q.jpg
skipimage: true 
 
excerpt: "This is the first blog post in a series of posts where we introduce using pretrained Transformer models for search and document ranking with Vespa.ai."  

---

![Decorative image](/assets/2021-05-19-pretrained-transformer-language-models-for-search-part-1/jamie-street-uNNCs5kL70Q.jpg)
<p class="image-credit"> 
Photo by <a href="https://unsplash.com/@jamie452?utm_source=unsplash&utm_medium=referral&utm_content=creditCopyText">Jamie Street</a> on <a href="https://unsplash.com/s/photos/retriever?utm_source=unsplash&utm_medium=referral&utm_content=creditCopyText">Unsplash</a></p>

> In this blog series we demonstrate how to represent transformer models in a multi phase retrieval and ranking pipeline using Vespa.ai. We also evaluate these models on the largest Information Retrieval relevance dataset, namely the MS Marco Passage ranking dataset. We demonstrate how to achieve close to state of the art ranking using miniature transformer models with just 22M parameters,  beating large ensemble models with billions of parameters.

Reviewer: ```suggestion
In this blog series we demonstrate how to represent transformer models in a multiphase retrieval and ranking pipeline using Vespa.ai. We also evaluate these models on the largest Information Retrieval relevance dataset, namely the MS Marco Passage ranking dataset. We demonstrate how to achieve close to state of the art ranking using miniature transformer models with just 22M parameters,  beating large ensemble models with billions of parameters.
```

>  In this first post we give an introduction to Transformers for text ranking and three different methods of applying them for ranking.  We also cover multi-phase retrieval and ranking pipelines and introduce three different ways to efficiently retrieve candidate documents in a multi-phase retrieval and ranking pipeline.

Reviewer: ```suggestion
 In this first post we give an introduction to Transformers for text ranking and three different methods of applying them for ranking.  We also cover multiphase retrieval and ranking pipelines and introduce three different ways to efficiently retrieve candidate documents in a multiphase retrieval and ranking pipeline.
```

![table](/assets/2021-05-19-pretrained-transformer-language-models-for-search-part-1/image1.png)

The table shows how significant the advancement was when first applied to the [MS MARCO Passage Ranking leaderboard](https://microsoft.github.io/MSMARCO-Passage-Ranking-Submissions/leaderboard/). The state of the art results on MS Marco passage ranking advanced by almost 30% within a week, while improvements up until then had been incremental at best. 
Compared to the baseline BM25 text ranking (default Apache Lucene 9 text scoring), applying BERT improved the ranking effectiveness by more than 100%. 

The table is from [Pretrained Transformers for Text Ranking: BERT and Beyond](https://arxiv.org/abs/2010.06467) which is a brilliant resource for understanding how pretrained transformers models can be used for document ranking. The MS MARCO Passage ranking relevancy dataset consists of about 8.8M passages and more than 500 000 queries with at least one judged relevant document. It is by far the largest IR dataset available in the public domain and is commonly used to evaluate ranking models. 

> The MS Marco passage ranking dataset queries are split in three different subsets, the train, development (dev) and test (eval). The train split can be used to train a ranking model using machine learning. Once a model is built, one can test the effectiveness of the ranking model on the development and test split. Applying the learned model on the development and test set is called in-domain usage of the model.  If the trained ranking model is applied on a different relevancy dataset, it’s usually referred to as out of domain usage, or zero-shot.How well models trained on MS Marco query and passage pairs generalize to other domains is out of scope for this blog post, but we can sincerely recommend [BEIR: A Heterogenous Benchmark for Zero-shot Evaluation of Information Retrieval Models](https://arxiv.org/abs/2104.08663). 

Reviewer: ```suggestion
The MS Marco passage ranking dataset queries are split in three different subsets, the train, development (dev) and test (eval). The train split can be used to train a ranking model using machine learning. Once a model is built, one can test the effectiveness of the ranking model on the development and test split. Applying the learned model on the development and test set is called in-domain usage of the model.  If the trained ranking model is applied on a different relevancy dataset, it’s usually referred to as out of domain usage, or zero-shot. How well models trained on MS Marco query and passage pairs generalize to other domains is out of scope for this blog post, but we can sincerely recommend [BEIR: A Heterogenous Benchmark for Zero-shot Evaluation of Information Retrieval Models](https://arxiv.org/abs/2104.08663). 
```

> The official evaluation metric used for the MS Marco Passage ranking leaderboard is MRR@10. The name might sound scary but it’s in reality a trivial way to judge the effectiveness of a text ranking algorithm. RR@10 is the Reciprocal Rank of the first relevant passage within the top 10 ranking positions for a given query.  **@k** denotes the depth into the top ranking documents we look for the relevant document(s). The reciprocal rank formula is simply 1/(position of the first relevant hit). If the judged relevant hit (as judged by a human) is ranked at position 1 the reciprocal rank score is 1. If the relevant hit is found at position 2 the reciprocal rank score is 0.5 and so on.The mean in mean reciprocal rank is simply the mean rr over all queries in the dev or test split which gives us a score for the overall ranking algorithm. The MS Marco passage ranking development (dev) set consists of 6,980 queries so it’s a rather large pool of queries.

Reviewer: ```suggestion
The official evaluation metric used for the MS Marco Passage ranking leaderboard is MRR@10. The name might sound scary but it’s in reality a trivial way to judge the effectiveness of a text ranking algorithm. RR@10 is the Reciprocal Rank of the first relevant passage within the top 10 ranking positions for a given query.  **@k** denotes the depth into the top ranking documents we look for the relevant document(s). The reciprocal rank formula is simply 1/(position of the first relevant hit). If the judged relevant hit (as judged by a human) is ranked at position 1 the reciprocal rank score is 1. If the relevant hit is found at position 2 the reciprocal rank score is 0.5 and so on. The mean in mean reciprocal rank is simply the mean rr over all queries in the dev or test split which gives us a score for the overall ranking algorithm. The MS Marco passage ranking development (dev) set consists of 6,980 queries so it’s a rather large pool of queries.
```

As demonstrated in [Pretrained Transformers for Text Ranking: BERT and Beyond](https://arxiv.org/abs/2010.06467), pre-trained language models of the Transformer family achieve best accuracy for text ranking and question answering tasks when used as an interaction model with all-to-all cross-attention between the query and document. Generally there are 3 ways to use Transformer models for text ranking and all of them require training data to fine tune for retrieval or ranking. 

![table](/assets/2021-05-19-pretrained-transformer-language-models-for-search-part-1/image3.png)

Figure from [ColBERT: Efficient and Effective Passage Search via Contextualized Late Interaction over BERT](https://arxiv.org/abs/2004.12832) illustrating various deep neural networks for ranking. In the following section we give an overview of three of these methods  using Transformer models. 

# Representation based ranking using Transformer models

> It is possible to use the Transformer model as the underlying deep neural network for representation based learning. Given training data, one can learn a representation of documents and queries so that relevant documents are closer or more similar in this representation than irrelevant documents.The representation based ranking approach falls under the broad representation learning research field and representation learning can be applied to text, images, videos or even combinations (multi-modal representations).  For text ranking, queries and documents are embedded into a dense vector embedding space using one or two Transformer models (bi-encoders).  The embedding representation is learned by the training examples given to the model. Once the model has been trained, one can pre-compute the embeddings for all documents and at run time use nearest neighbor search for efficient document retrieval or (re)ranking. Documents are typically ranked by the distance to the query representation in the embedding vector space or by maximum dense inner product search.  

Reviewer: ```suggestion
It is possible to use the Transformer model as the underlying deep neural network for representation based learning. Given training data, one can learn a representation of documents and queries so that relevant documents are closer or more similar in this representation than irrelevant documents. The representation based ranking approach falls under the broad representation learning research field and representation learning can be applied to text, images, videos or even combinations (multi-modal representations).  For text ranking, queries and documents are embedded into a dense vector embedding space using one or two Transformer models (bi-encoders).  The embedding representation is learned by the training examples given to the model. Once the model has been trained, one can pre-compute the embeddings for all documents and at run time use nearest neighbor search for efficient document retrieval or (re)ranking. Documents are typically ranked by the distance to the query representation in the embedding vector space or by maximum dense inner product search.  
```

Similar to the pure representation based model we only need to encode the query through the transformer model at query time. The query tokens only attend to other query tokens, and similar document tokens only attend to other document tokens.  

As demonstrated in the ColBERT paper, the late interaction ColBERT model achieves almost the same ranking accuracy as the more computationally complex all to all query document interaction models. The model is trained in the same way as with representation or all to all interaction models. The downside of the late interaction model is the storage cost of the document term embeddings. Instead of one embedding like with the representation model there is an embedding vector per term in the document.  How large depends on the number of dimensions and the precision used per value (e.g. using float16 saves 50% compared to float32). 

> # Multi phase retrieval and ranking 

Reviewer: ```suggestion
# Multiphase retrieval and ranking 
```

> Due to computationally complexity of especially the all to all interaction model there has been renewed interest in multi phase retrieval and ranking. In a multi-phased retrieval and ranking pipeline the first phase retrieves candidate documents using a cost efficient retrieval method and the more computationally complex cross-attention or late interaction model inference is limited to the top ranking documents from the first phase. 

Reviewer: ```suggestion
Due to computationally complexity of especially the all to all interaction model there has been renewed interest in multiphase retrieval and ranking. In a multiphased retrieval and ranking pipeline, the first phase retrieves candidate documents using a cost efficient retrieval method and the more computationally complex cross-attention or late interaction model inference is limited to the top ranking documents from the first phase. 
```

## Dense Retrieval
Embedding based models embed or map queries and documents into a latent low dimensional dense embedding vector space and use vector search to retrieve documents. Dense retrieval could be accelerated by using approximate nearest neighbor search, for example indexing the document vector representation using [HNSW](https://docs.vespa.ai/en/approximate-nn-hnsw.html) graph indexing. In-domain dense retrievers based on bi-encoder architecture trained on MS Marco passage data have demonstrated that they can outperform sparse lexical retrievers with a very large margin. In a multi stage retriever and ranking pipeline, the first stage is focused on getting a good recall at k where k is the re-ranking depth we can afford from a computationally complexity perspective (cost and serving latency). 

> Both of these retrieval methods are capable of searching through millions of documents per node with very low latency and cost and Vespa supports both methods. 

Reviewer: ```suggestion
Both of these retrieval methods are capable of searching through millions of documents per node with very low latency and cost, Vespa supports both methods. 
```
